### ISSUE TITLE
Avoiding storage waste when data is already replicated at an application-level, while maintaining high-availability

### ISSUE THREAD

**Issue by michaelgeorgeattard** [rook/rook#9824]
We have multiple HA applications that replicate their data across members (ex. Consul, Kafka, ETCD etc.)

As an example, let's take a cluster of 3 ETCD members storing data in a CephFS PVC which in turn has 3 replicas of the data stored in Ceph.

In this case 1GB of actual data would take 1X3X3 = 9GB of disk space.

If we tweak the CephFS storage class to have 1 replica, we would suffer downtime during K8s cluster maintenance.

Apart from using Erasure Coding to reduce the amount of storage needed, are there any other options that we can leverage for data that is already replicated at an application level?

**Comment by travisn** [comment]
If the application has its own replication, it’s likely preferred for the application to use local storage. The downside is of course you lose portability of the application to other nodes.

Ceph really doesn’t handle replica 1 well. Since the data is spread across all the OSDs, the outage of any OSD would cause the data to become unavailable, so the min recommendation is replica 2 or erasure coding as you mention.

**Comment by spdfnet** [comment]
This would be better classified as an anti-pattern, not waste.

**Comment by michaelgeorgeattard** [comment]
Given the example above with 3 ETCD pods, each bound to a different `PVC` backed using a `CephFilesystem` with a replication factor of 1.

Would it be possible to have that replica for each PVC on a different Ceph host? The host could be where the ETCD pod is running or otherwise.

Not sure if the above is an anti-pattern, but it might need rephrasing due to my limited knowledge on Rook and Ceph in general.

Our Kubernetes clusters are on-premise, and PVCs backed by local storage would bring a lot of limitions.

**Comment by spdfnet** [comment]
What I meant was, given the ETCD example, each member of the cluster expects its storage to be isolated.
If there is a Ceph issue, the whole ETCD cluster would be down.
That is more than just a waste of space, it is also a way to compromise the distributed nature of ETCD.

**Comment by michaelgeorgeattard** [comment]
I found https://github.com/rook/rook/issues/2959, please close this issue if it doesn't add any value.

**Comment by michaelgeorgeattard** [comment]
> What I meant was, given the ETCD example, each member of the cluster expects its storage to be isolated. If there is a Ceph issue, the whole ETCD cluster would be down. That is more than just a waste of space, it is also a way to compromise the distributed nature of ETCD.

Partial Ceph failures (1 host is down) wouldn't have an impact on ETCD if we are using a replication factor of 3 for example no?

**Comment by spdfnet** [comment]
There are multiple scenarios where the primary OSD of a Placement Group would have a byzantine failure, and block writes to that PG.
For linear/atomic systems like ETCD, Postgres, etc. it means loss of service, even if 99% of data is active+clean in Ceph.

**Comment by travisn** [comment]
> I found #2959, please close this issue if it doesn't add any value.

Even if you can configure the storage to run on particular nodes by using the deviceClass or other properties, Ceph still requires at least replica 2. By nature, Ceph is a distributed storage system and does not provide local node affinity. What are you trying to gain with Ceph instead of a local volume in this case?

**Comment by michaelgeorgeattard** [comment]
> > I found #2959, please close this issue if it doesn't add any value.
> 
> Even if you can configure the storage to run on particular nodes by using the deviceClass or other properties, Ceph still requires at least replica 2. By nature, Ceph is a distributed storage system and does not provide local node affinity. What are you trying to gain with Ceph instead of a local volume in this case?

The ability to use CSI snapshots for example, however you might raise the point that snapshots might not be the best way to backup a workload that is replicated at an application-level.

Another advantage is the ability to schedule the workload pod on different nodes if needed.

Finally we have the ability to Ceph mirroring to replicate data outside of the cluster for DRP, we would need to have a separate solution for local storage.

Basically the best counter against using local storage would be it would be ideal to use the same storage solution (Ceph) for all workloads.

I would have assumed that workloads that are replicated at an application-level (referred to business level in https://github.com/rook/rook/issues/2959) are a very common scenario, and I would naively assume that Ceph + Rook would cater for such scenarios.

**Comment by travisn** [comment]
If you need all that functionality, then Ceph is certainly going to be the right platform, but it really does require its own replication, which means duplicate storage for the applications that have their own replication. 

Applications that have their own replication also typically want to have the higher performance of a local disk, which Ceph will not be able to keep up with since there is a network in the middle.
